functions/anonymous-functions.md

```diff
@@ -0,0 +1,198 @@
+# Anonymous Functions
+
+## Introduction
+
+Yet another permutation of the function in the JavaScript language!
+
+Anonymous functions are a useful way to define functions in the context they're used. This usage can be more readable sometimes, especially when it's more convenient to read a function's definition immediately over chase its definition, which may be in a different file.
+
+## Syntax
+
+An anonymous function is a function created with any valid syntax, but it is not referenced by a variable.
+
+Instead, an **anonymous function** is a function defined in the one place that it's used.
+
+Technically, many of the methods in objects we created are anonymous functions!
+
+<!-- prettier-ignore-start -->
+```js
+const myTedTalk = {
+  title: 'Inspiration From Inspiring Things',
+  summary: 'A story about where inspiration comes from.',
+  summarize: function() {
+    return `${this.title}: ${this.summary}`;
+  }
+}
+
+myTedTalk.summarize();
+```
+<!-- prettier-ignore-end -->
+
+We never _named_ the value of the `summarize` property, or referenced it in a variable. We defined the function in the one place where it's needed, anonymously!
```

Comment:
Strongly suggest avoiding dragging in one-shot object functions here (I voiced my strong opposition to these being in the curriculum at all in methods in objects), but I do think we need to build up to what we mean by anonymous function, since it's not really clear from the typical syntax we'll be using to create functions whether they have a name, or what effect that has (spoilers, no effect). What we're calling anonymous functions are really  (dun dun duuun) one-shot functions/lambda/closures/some combination of function buzzwords. I can get on board with calling them anonymous functions, but in that case, we especially need to give some development to get from the intuitive meaning of the word anonymous to what we actually care about for the lesson.

Also included a callout tying these functions back to callback functions.

So I would replace this content with the markdown in the following comment.
This suggestion just has the removal of this content.
Could not use a suggestion for the new content due to code snippets.

```suggestion
```

Reply:
Technically, any arrow function or function expression _could_ be considered an anonymous function. In either situation, we are creating a function, a reference to which is stored in a variable.

Even though the expression does not include a name as part of the function definition, the JavaScript runtime recognizes the pattern of assigning these functions to a variable and will take the variable name as the _implicit_ name of the function object.

Consider these examples:

<!-- prettier-ignore-start -->
```js
function nameChecker(fn) {
  console.log(fn.name);  
}

// arrow function
const myfn = () => 0;
nameChecker(myfn);  // myfn

// function expression
const yourfn = function() { return 0; };
nameChecker(yourfn);  // yourfn

// and for completeness...
nameChecker(nameChecker);  // nameChecker
```
<!-- prettier-ignore-end -->

But what about this?

<!-- prettier-ignore-start -->
```js
function nameChecker(fn) {
  console.log(fn.name);  
}

nameChecker(() => 0);  // blank!
nameChecker(function() { return 0; });  // also blank!
```
<!-- prettier-ignore-end -->

This time, the functions have empty names. More over, while these functions can be accessed inside the function into which they were passed, once that function returns we no longer have any reference to that function! Each function was created exactly where it was intended to be used, and afterwards it effectively disappears!

An **anonymous function** is a function created with any valid syntax, defined in the one place that it's used.

### !callout-info

## A Function Is a Function Is a Function

Whether or not a function has a name, or gets stored in a variable for later use, a function is still a function. There's nothing special about passing an anonymous function as a callback. Anywhere that we pass an anonymous function that we create right when we need, we could pass a named function that we created elsewhere.

<br />

As we will see, using an anonymous function, a function we created right where we needed it, can simply be more convenient.

### !end-callout

Reply:
I'm into the direction of this, I'm inclined to simplify it more. Also, I think that this section should still start with a strong definition of "anonymous functions"

---

An **anonymous function** is a function created with any valid syntax, defined in the one place that it's used.

Every function has a property named `name`, which holds the name of the function. We can access it through dot notation. For example, we can create a function named `helloWorldFn`:

```js
const helloWorldFn = function() {
    console.log('hello world!');
}
```

When we print this function's `name`, we'll get `helloWorldFn` back.

```js
console.log('the name of the function helloWorldFn is:', helloWorldFn.name);
```

Output:

```
the name of the function helloWorldFn is: helloWorldFn
```

An anonymous function is defined in the one place that it's used. These functions do not have a defined `name` property.

For example, let's print the `name` of an anonymous function in the one place it's used:

```js
console.log('The name of this anonymous function is blank!:', (() => 0).name);
```

Here, our one-line arrow function `() => 0` returns zero. We wrapped it with parentheses `()` and then accessed its `name` property. We get this output:

```
The name of this anonymous function is blank!:
```

Each function was created exactly where it was intended to be used, and afterwards it effectively disappears!


### !callout-info

## A Function Is a Function Is a Function

Whether or not a function has a name, or gets stored in a variable for later use, a function is still a function. There's nothing special about passing an anonymous function as a callback. Anywhere that we pass an anonymous function that we create right when we need, we could pass a named function that we created elsewhere.

<br />

As we will see, using an anonymous function, a function we created right where we needed it, can simply be more convenient.

### !end-callout